### PR TITLE
[mysql] Support tables which do not contain a primary key (#1827)

### PR DESCRIPTION
**Pull request by ruanhang1993**
This PR implements #1827.

### CODE REVIEW

**Review by zstraw — COMMENTED** [review]
Thanks for pushing this issue forward. 

NIT: The doc has not updated.

About the [question](https://github.com/ververica/flink-cdc-connectors/pull/2046#discussion_r1169757507) I mentioned before,  shall we provide users with the option of exactly-once but worse performance if no pk table?

File: flink-connector-mysql-cdc/src/main/java/com/ververica/cdc/connectors/mysql/source/config/MySqlSourceConfigFactory.java [hunk 1]
```diff
@@ -72,6 +73,7 @@ public class MySqlSourceConfigFactory implements Serializable {
     private Duration heartbeatInterval = HEARTBEAT_INTERVAL.defaultValue();
     private Properties dbzProperties;
     private String chunkKeyColumn;
+    private List<String> specificPks = new ArrayList<>();
```
Comment (leonardBang) [inline flink-connector-mysql-cdc/src/main/java/com/ververica/cdc/connectors/mysql/source/config/MySqlSourceConfigFactory.java:76]: Could current design cover the case which user captures two NoPKTable with different speciffic primary keys(talbe 1: cola,colb, table 2：col1, col2)?

**Review by leonardBang — COMMENTED** [review]
Thanks @ruanhang1993  for the contribution, I left some comments.

File: flink-connector-mysql-cdc/src/main/java/com/ververica/cdc/connectors/mysql/debezium/reader/SnapshotSplitReader.java [hunk 2]
```diff
@@ -403,4 +398,77 @@ public boolean isRunning() {
...
+                return splitKeyRangeContains(
+                        key,
+                        currentSnapshotSplit.getSplitStart(),
+                        currentSnapshotSplit.getSplitEnd());
+            }
+            return false;
+        }
+    }
+
+    /** Collect records without primary key. There are no binlog events. */
```
[the remark this answers is not in the document]
Comment (ruanhang1993) [reply flink-connector-mysql-cdc/src/main/java/com/ververica/cdc/connectors/mysql/debezium/reader/SnapshotSplitReader.java:443]: Forget change its message. Fixed.

File: flink-connector-mysql-cdc/src/main/java/com/ververica/cdc/connectors/mysql/table/MySqlTableSourceFactory.java [hunk 3]
```diff
@@ -267,12 +271,14 @@ private static StartupOptions getSpecificOffset(ReadableConfig config) {
...
-    private void validatePrimaryKeyIfEnableParallel(ResolvedSchema physicalSchema) {
-        if (!physicalSchema.getPrimaryKey().isPresent()) {
+    private void validatePrimaryKeyIfEnableParallel(
+            ResolvedSchema physicalSchema, @Nullable String chunkKeyColumn) {
+        if (chunkKeyColumn == null && !physicalSchema.getPrimaryKey().isPresent()) {
             throw new ValidationException(
                     String.format(
-                            "The primary key is necessary when enable '%s' to 'true'",
-                            SCAN_INCREMENTAL_SNAPSHOT_ENABLED));
+                            "The primary key is necessary when enable '%s' to 'true'. If the table doesn't have primary keys, '%s' must be set.",
```
Comment (leonardBang) [inline flink-connector-mysql-cdc/src/main/java/com/ververica/cdc/connectors/mysql/table/MySqlTableSourceFactory.java:279]: '%s' is required for table without primary key when  '%s'  enabled .

File: flink-connector-mysql-cdc/src/main/java/com/ververica/cdc/connectors/mysql/debezium/reader/BinlogSplitReader.java [hunk 4]
```diff
@@ -253,6 +256,20 @@ private boolean shouldEmit(SourceRecord sourceRecord) {
...
     }
 
+    private Struct getStructContainsChunkKey(SourceRecord record) {
+        if (record.key() != null) {
+            return (Struct) record.key();
+        }
+
+        Envelope.Operation op = Envelope.operationFor(record);
+        Struct value = (Struct) record.value();
+        if (op == Envelope.Operation.CREATE || op == Envelope.Operation.READ) {
```
Comment (leonardBang) [inline flink-connector-mysql-cdc/src/main/java/com/ververica/cdc/connectors/mysql/debezium/reader/BinlogSplitReader.java:266]: please add some notes for the two cases

File: flink-connector-mysql-cdc/src/main/java/com/ververica/cdc/connectors/mysql/source/utils/RecordUtils.java [hunk 5]
```diff
@@ -113,6 +114,121 @@ public static void upsertBinlog(
...
+
+    private static void upsertBinlogForNoPKTable(
+            List<SourceRecord> snapshotRecords,
+            SourceRecord binlogRecord,
+            Struct value,
+            boolean isDelete) {
+        if (isDelete) {
+            Iterator<SourceRecord> iterator = snapshotRecords.iterator();
+            while (iterator.hasNext()) {
+                SourceRecord sourceRecord = iterator.next();
```
Comment (leonardBang) [inline flink-connector-mysql-cdc/src/main/java/com/ververica/cdc/connectors/mysql/source/utils/RecordUtils.java:201]: Iterating is slow with the records number growing, could we use `map<row, list<row>>` to improve it ?

Comment (leonardBang) [reply flink-connector-mysql-cdc/src/main/java/com/ververica/cdc/connectors/mysql/source/utils/RecordUtils.java:201]: and in this way, we can reduce `upsertBinlogForNoPKTable` and existed `upsertBinlog` to one method.

File: flink-connector-mysql-cdc/src/main/java/com/ververica/cdc/connectors/mysql/source/assigners/MySqlChunkSplitter.java [hunk 6]
```diff
@@ -146,7 +146,9 @@ private void analyzeTable(TableId tableId) {
             currentSplittingTable = mySqlSchema.getTableSchema(jdbcConnection, tableId).getTable();
             splitColumn =
                     ChunkUtils.getChunkKeyColumn(
-                            currentSplittingTable, sourceConfig.getChunkKeyColumn());
+                            currentSplittingTable,
+                            sourceConfig.getChunkKeyColumn(),
+                            sourceConfig.getSpecificPks());
```
[the remark this answers is not in the document]
Comment (ruanhang1993) [reply flink-connector-mysql-cdc/src/main/java/com/ververica/cdc/connectors/mysql/source/assigners/MySqlChunkSplitter.java]: This option is reverted.

File: flink-connector-mysql-cdc/src/main/java/com/ververica/cdc/connectors/mysql/source/utils/RecordUtils.java [hunk 7]
```diff
@@ -73,46 +79,150 @@ public static Object[] rowToArray(ResultSet rs, int size) throws SQLException {
...
+            Map<Struct, List<SourceRecord>> snapshotRecords,
+            SourceRecord binlogRecord,
+            Struct keyStruct,
+            boolean isDelete) {
+        boolean hasPrimaryKey = binlogRecord.key() != null;
+        List<SourceRecord> records = snapshotRecords.get(keyStruct);
+        if (isDelete) {
+            if (records == null || records.isEmpty()) {
+                LOG.error(
+                        "Deleting a record which is not in its split for tables without primary keys. This may happen when the chunk key column is updated in another snapshot split.");
```
Comment (leonardBang) [inline flink-connector-mysql-cdc/src/main/java/com/ververica/cdc/connectors/mysql/source/utils/RecordUtils.java:187]: we need open a doc PR to guide user how to choose chunk key column

**Review by leonardBang — APPROVED** [review]
Thanks @ruanhang1993 for the update, LGTM

**Comment by ruanhang1993** [comment]
> Thanks for pushing this issue forward.
> 
> NIT: The doc has not updated.
> 
> About the [question](https://github.com/ververica/flink-cdc-connectors/pull/2046#discussion_r1169757507) I mentioned before, shall we provide users with the option of exactly-once but worse performance if no pk table?

Hi, @zstraw .
The single split is actually hard to use for users. All table data will be read into the memory and may cause OOM.
So this implementation requires the chunk key column when the table does not have primary keys.

I will provide the docs about these experimental features.

File: flink-connector-mysql-cdc/src/main/java/com/ververica/cdc/connectors/mysql/source/utils/RecordUtils.java (hunk 7 above)
Comment (ruanhang1993) [reply flink-connector-mysql-cdc/src/main/java/com/ververica/cdc/connectors/mysql/source/utils/RecordUtils.java:187]: I will provide the docs about these experimental features later. See #2164.

**Pull request merged**